fix: user catalog

axel7083 opened the pull request:
### What does this PR do?

See https://github.com/containers/podman-desktop-extension-ai-lab/issues/792#issuecomment-2060792927 for details

When we were importing a model we were copying to existing catalog, and creating a new one with the user's models in it. This lead to the issue that the catalog was not updated.

This PR fixes this issue, by having the user's catalog a complementary catalog, being merged at startup with the one provided by the extension. By this method, the user can define recipes, models or categories in a json file, and they will be added to the existing one if properly formatted.

When importing models, this file is automatically created and filled with the user's models details.

### Screenshot / video of UI

https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/297da5c2-78e6-48dd-95db-cdf2affb1b99

### What issues does this PR fix or reference?

Fixes https://github.com/containers/podman-desktop-extension-ai-lab/issues/792

### How to test this PR?

- [x] existing tests ensure no regression.

#### Manually

- (1) Go to the models catalog and import models
- (2) Delete imported models
- (3) Restart AI-Lab
- (4) Delete the `user-catalog.json` file

jeffmaury reviewed (changes requested):
README.Md should be updaed as well as the file name changed

axel7083 commented on packages/backend/src/managers/catalogManager.ts:
```diff
@@ -51,31 +53,88 @@ export class CatalogManager extends Publisher<ApplicationCatalog> implements Dis
+
+    const sanitize = this.sanitize(content);
+    this.catalog = {
+      models: [...defaultCatalog.models, ...sanitize.models],
```
He can't. What would be the use case ?

lstocchi replied:
I think we should check if there are multiple items with the same id. In the rare case that it may happen we could have issues as we manage the models as a map.

lstocchi replied:
If that's the goal it already works this way. If you have in the user catalog an entry with an id = one item in the default catalog you will see the custom one

axel7083 commented:
@jeffmaury @benoitf 

https://github.com/containers/podman-desktop-extension-ai-lab/blob/c97f233e9a4487148c793e84e624f4e88f094baf/README.md?plain=1#L134

I don't know what to replace this path with, as we are using the podman desktop api storage, this path could be different depending on the installation.

lstocchi reviewed (commented):
Works fine for me. There is just the edge case when there are two items having the same id to be decided how to be handled

lstocchi commented on README.md:
```diff
@@ -129,15 +129,15 @@ For the time being, please consider the following actions:
+The extension provides by default a curated list of recipes, models and categories. However, this system is extensible and you can define your own.
 
-The catalog provides lists of categories, recipes, and models.
+To enhance the existing catalog, you can create a file located in the extension storage folder `$HOME/.local/share/containers/podman-desktop/extensions-storage/redhat.ai-lab/user-catalog.json` in the extension storage folder.
```
You repeated twice `in the extension storage folder`. One of the two can be removed

axel7083 replied:
Thanks, fixed

jeffmaury commented:
> Works fine for me. There is just the edge case when there are two items having the same id to be decided how to be handled

I would be to give preference to the user one but that could be a follow up PR

lstocchi commented on README.md:
```diff
@@ -129,15 +129,15 @@ For the time being, please consider the following actions:
+To enhance the existing catalog, you can create a file located in the extension storage folder `$HOME/.local/share/containers/podman-desktop/extensions-storage/redhat.ai-lab/user-catalog.json` in the extension storage folder.
 
-Each recipe can belong to one or several categories. Each model can be used by one or several recipes.
+It must follow the same format that the default catalog [in the sources of the extension](https://github.com/containers/podman-desktop-extension-ai-lab/blob/main/packages/backend/src/assets/ai.json). 
```
```suggestion
It must follow the same format as the default catalog [in the sources of the extension](https://github.com/containers/podman-desktop-extension-ai-lab/blob/main/packages/backend/src/assets/ai.json). 
```

axel7083 commented:
@lstocchi @jeffmaury 

I updated the code to let the user's item overwrite the default items when ids are conflicting.

![image](https://github.com/containers/podman-desktop-extension-ai-lab/assets/42176370/1f58ff93-066f-4e19-aa62-3119fb2a5e17)

> ℹ️  Here is an example overwriting the ChatBot.

I also added two tests to ensure the expected behaviour.

lstocchi reviewed (approved):
LGTM 🚀

jeffmaury reviewed (approved):
LGTM
